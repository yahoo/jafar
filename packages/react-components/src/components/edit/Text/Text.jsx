/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import Input from '@material-ui/core/Input';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';

/**
 * Represent a string value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Text/Text.jsx">
 Text</a> from '@jafar/react-components/edit/Text'
 */
export default class Text extends React.Component {
  static propTypes = {
    value: PropTypes.string,
    state: PropTypes.shape({
      placeholder: PropTypes.string,
      multiline: PropTypes.bool,
      rowsMax: PropTypes.number,
      maxLength: PropTypes.number,
    }),
    disabled: PropTypes.bool,
    invalid: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: '',
    state: {
      placeholder: '',
      multiline: false,
      rowsMax: undefined,
      maxLength: undefined,
    },
    disabled: false,
    invalid: false,
  };

  render() {
    return (
      <FormControl>
        <Input
          type="text"
          value={this.props.value}
          placeholder={this.props.state.placeholder}
          disabled={this.props.disabled}
          multiline={this.props.state.multiline}
          rowsMax={this.props.state.rowsMax}
          error={this.props.invalid}
          onChange={this.onValueChange}
          inputProps={{
            maxLength: this.props.state.maxLength,
          }} />
        { this.props.state.maxLength === undefined ? null :
          <FormHelperText>{this.props.value.length} / {this.props.state.maxLength}</FormHelperText> }
      </FormControl>
    );
  }

  onValueChange = (e) => {
    this.props.onValueChange(e.target.value);
  }
}
