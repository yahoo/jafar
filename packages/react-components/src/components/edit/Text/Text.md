<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    firstName: {
      label: 'First Name',
      path: 'firstName',
      component: {
        name: 'Text',
        state: {
            maxLength: 30,
        }
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      firstName: 'Rachel', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    Text: {
        renderer: Text,
    }, 
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='firstName' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 'Rachel Green',
    state: {
        placeholder: 'Enter Name',
    },
};

 <Text
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Multiline</h4>

```javascript
initialState = { 
    value: `Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.`,
    state: {
        placeholder: 'Enter Name',
        multiline: true,
        rowsMax:10,
    },
};
 <Text
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    value: undefined,
    state: {
        placeholder: 'Enter Name',
    },
};
 <Text
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Disabled</h4>

```javascript
initialState = { 
    value: 'Rachel Green',
    state: {
        placeholder: 'Enter Name',
    },
};
 <Text
    value={state.value}
    state={state.state}
    disabled={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Invalid</h4>

```javascript
initialState = { 
    value: 'Rachel Green',
    state: {
        placeholder: 'Enter Name',
    },
};
 <Text
    value={state.value}
    state={state.state}
    invalid={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```