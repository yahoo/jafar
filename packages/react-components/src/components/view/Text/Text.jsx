/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';

/**
 * Represent a string value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/view/Text/Text.jsx">
 Text</a> from '@jafar/react-components/view/Text'
 */
export default class Text extends React.Component {
  static propTypes = {
    value: PropTypes.string,
  };

  static defaultProps = {
    value: '',
  };

  render() {
    return (<div>{this.props.value}</div>);
  }
}
