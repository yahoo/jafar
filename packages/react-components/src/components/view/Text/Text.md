<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    firstName: {
      label: 'First Name',
      path: 'firstName',
      component: {
        name: 'Text',
      }
    },
  },
  data: {
      firstName: 'Rachel', 
  },
};

const resources = {
  components: { 
    Text: {
      renderer: Text,
    }, 
  }
};

<Form model={model} resources={resources}>
    <Field id='firstName' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 'Rachel Green'
};

<Text value={state.value} />
```
