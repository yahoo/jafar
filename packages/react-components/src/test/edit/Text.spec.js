/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Text from '../../components/edit/Text/index';

describe('<Text />', () => {
  let component;
  let onValueChangeSpy;
  const value = 'avocado';
  const state = { placeholder: 'Enter Food' };
  const disabled = false;
  const invalid = false;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onValueChange callback on value change', () => {
    component = mount(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    const newValue = 'banana';
    component.find('input').simulate('change', { target: { value: newValue } });
    expect(onValueChangeSpy).toHaveBeenCalledWith(newValue);
  });

  function getComponent(value, state, disabled, invalid, onValueChange) {
    return (
      <Text
        value={value}
        state={state}
        disabled={disabled}
        invalid={invalid}
        onValueChange={onValueChange}
      />
    );
  }
});
