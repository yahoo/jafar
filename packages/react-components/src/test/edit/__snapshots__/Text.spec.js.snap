// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should render provided data 1`] = `
<WithStyles(ForwardRef(FormControl))>
  <WithStyles(ForwardRef(Input))
    disabled={false}
    error={false}
    inputProps={
      Object {
        "maxLength": undefined,
      }
    }
    onChange={[Function]}
    placeholder="Enter Food"
    type="text"
    value="avocado"
  />
</WithStyles(ForwardRef(FormControl))>
`;
